### PR TITLE
Poll Plugin

### PR DESCRIPTION
**Pull request by 1999-XiChi**
+ feature
  + 添加了i18n（日文是谷歌翻译转换的，不确保是否正确）
  + 新增依赖date-fns库，格式化日期
  + 使用原有gun实例，抽象简化了gundb的代码
+ fix
  + 修复了pollInPost函数的重复渲染问题，投票后等待时间很短
+ todo
  + 智能合约solidity，接入以太坊
  + 丰富投票的类型（多选，限制人数...）
  + 安全性方面

### CODE REVIEW

File: src/extension/background-script/PluginService.ts [hunk 1]
```diff
@@ -14,6 +15,7 @@ const Plugins = {
     'maskbook.fileservice': FileService,
     'maskbook.trader': Trader,
     'co.gitcoin': Gitcoin,
+    'maskbook.polls': Poll,
```
Comment (septs) [inline src/extension/background-script/PluginService.ts:18]: move to `maskbook.trader` below.

File: src/plugins/Polls/Services.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    options: Array<string>
+    start_time: Date
+    end_time: Date
+}
+
+export async function createNewPoll(poll: NewPollProps) {
+    const { id, options, start_time, end_time } = poll
+
+    const results = new Array<number>(options.length).fill(0)
+    const resultsObj: Object = { ...results }
```
Comment (septs) [inline src/plugins/Polls/Services.ts:34]: plz do not use `Object` type.

File: src/plugins/Polls/Services.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+export async function getPollByKey(props: { key: string }) {
+    const keys = props.key.split('_')
+    let poll: PollGunDB = {
+        ...defaultPoll,
+        key: props.key,
+        id: first(keys),
+    }
+
+    PollGun
+        // @ts-ignore
```
Comment (septs) [inline src/plugins/Polls/Services.ts:106]: why `@ts-ignore`?

File: src/plugins/Polls/UI/Polls.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                                    width: `${(poll.results[index] / totalVotes) * 100}%`,
+                                }}></div>
+                            <div className={classes.text}>{option}</div>
+                        </div>
+                        <div className={classes.text}>{poll.results[index]}</div>
+                    </div>
+                ))}
+            </div>
+            <Typography variant="body2" classes={{ root: classes.deadline }}>
+                {isClosed ? 'Closed' : `${getDeadline(poll.end_time)}`}
```
Comment (septs) [inline src/plugins/Polls/UI/Polls.tsx:124]: i18n

File: src/plugins/Polls/UI/Polls.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    const isClosed = Date.now() > poll.end_time ? true : false
+
+    const totalVotes = poll.results.reduce(
+        (accumulator: number, currentValue: number): number => accumulator + currentValue,
+    )
+
+    const getDeadline = (date: number) => {
+        const deadline = new Date(date)
+        if (isValid(deadline)) {
+            return `Deadline: ${format(deadline, 'yyyy-MM-dd HH:mm:ss')}`
```
Comment (septs) [inline src/plugins/Polls/UI/Polls.tsx:83]: i18n

File: src/plugins/Polls/UI/PollsDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    senderName: string | undefined
+    senderFingerprint: string | undefined
+    switchToCreateNewPoll: () => void
+}
+
+function NewPollUI(props: PollsDialogProps & NewPollProps) {
+    const classes = useStylesExtends(useNewPollStyles(), props)
+    const [loading, setLoading] = props.loading
+    const [question, setQuestion] = useState('')
+    const [options, setOptions] = useState<Array<string>>(['', ''])
```
Comment (septs) [inline src/plugins/Polls/UI/PollsDialog.tsx:83]:
```suggestion
    const [options, setOptions] = useState<[string, string]>(['', ''])
```

File: src/plugins/Polls/UI/PollsDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    }
+
+    const addNewOption = () => {
+        setOptions(options.concat(['']))
+    }
+
+    const sendPoll = async () => {
+        const start_time = new Date()
+        const end_time = new Date(
+            start_time.getTime() + days * 24 * 60 * 60 * 1000 + hours * 60 * 60 * 1000 + minutes * 60 * 1000,
```
Comment (septs) [inline src/plugins/Polls/UI/PollsDialog.tsx:102]: https://date-fns.org/v2.16.1/docs/addDays
https://date-fns.org/v2.16.1/docs/addHours
https://date-fns.org/v2.16.1/docs/addMinutes

File: src/plugins/Polls/UI/PollsDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+    const { t } = useI18N()
+
+    const createNewPoll = () => {
+        setTabState(1)
+    }
+
+    const insertPoll = (data?: PollMetaData | null) => {
+        const ref = getActivatedUI().typedMessageMetadata
+        const next = new Map(ref.value)
+        data ? next.set('com.maskbook.poll:1', data) : next.delete('com.maskbook.poll:1')
```
Comment (septs) [inline src/plugins/Polls/UI/PollsDialog.tsx:283]: the key from constants file.

File: src/plugins/Polls/UI/PollsDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+        props.onConfirm()
+    }
+
+    const senderName = useCurrentIdentity()?.linkedPersona?.nickname
+    const senderFingerprint = useCurrentIdentity()?.linkedPersona?.fingerprint
+
+    const tabProps: AbstractTabProps = {
+        tabs: [
+            {
+                label: 'Create New',
```
Comment (septs) [inline src/plugins/Polls/UI/PollsDialog.tsx:294]: i18n

File: src/plugins/Polls/UI/PollsDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+                        loading={loading}
+                        switchToCreateNewPoll={createNewPoll}
+                        senderName={senderName}
+                        senderFingerprint={senderFingerprint}
+                    />
+                ),
+                p: 0,
+            },
+            {
+                label: 'Select Existing',
```
Comment (septs) [inline src/plugins/Polls/UI/PollsDialog.tsx:307]: i18n

File: src/plugins/Polls/types.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+export interface PollMetaData {
+    question: string
+    key: string
+    start_time: number
+    end_time: number
+    options: Array<string>
```
Comment (septs) [inline src/plugins/Polls/types.ts:6]:
```suggestion
    options: string[]
```

File: src/plugins/Polls/types.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+    results: Array<number>
```
Comment (septs) [inline src/plugins/Polls/types.ts:7]:
```suggestion
    results: number[]
```

File: src/plugins/Polls/utils.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+import { createTypedMessageMetadataReader, createRenderWithMetadata } from '../../protocols/typed-message/metadata'
+import type { PollMetaData } from './types'
+
+export const PollMetadataReader = createTypedMessageMetadataReader<PollMetaData>('com.maskbook.poll:1', {})
```
Comment (septs) [inline src/plugins/Polls/utils.ts:4]: ???

File: src/plugins/Polls/define.tsx [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+export const PollsPluginDefine: PluginConfig = {
+    pluginName: 'Poll',
+    identifier: 'com.maskbook.poll',
+    successDecryptionInspector: function Comp(props) {
+        const metadata = PollMetadataReader(props.message.meta)
+        if (!metadata.ok) return null
+        return <PollsInPost {...props} />
+    },
+    postDialogMetadataBadge: new Map([
+        ['com.maskbook.poll:1', (meta: PollMetaData) => `a poll of '${meta.question}' from ${meta.sender}`],
```
Comment (septs) [inline src/plugins/Polls/define.tsx:16]: ???

File: src/plugins/Polls/Services.ts [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+            poll = {
+                ...poll,
+                sender: data.sender,
+                question: data.question,
+                start_time: data.start_time,
+                end_time: data.end_time,
+            }
+            if (data.options) {
+                PollGun
+                    // @ts-ignore
```
Comment (septs) [inline src/plugins/Polls/Services.ts:118]: why `@ts-ignore` ?

File: src/plugins/Polls/Services.ts [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+                    .get(props.key)
+                    .get('options')
+                    .on((options) => {
+                        // @ts-ignore
```
Comment (septs) [inline src/plugins/Polls/Services.ts:122]: why `@ts-ignore` ?

File: src/plugins/plugin.ts [hunk 17]
```diff
@@ -23,12 +23,14 @@ export const PluginUI: ReadonlySet<PluginConfig> = plugins
...
 import { GitcoinPluginDefine } from './Gitcoin/define'
 import { RedPacketPluginDefine } from './RedPacket/define'
+import { PollsPluginDefine } from './Polls/define'
 import { StorybookPluginDefine } from './Storybook/define'
 import { FileServicePluginDefine } from './FileService/define'
 import { TraderPluginDefine } from './Trader/define'
 import { Flags } from '../utils/flags'
 plugins.add(GitcoinPluginDefine)
 plugins.add(RedPacketPluginDefine)
+plugins.add(PollsPluginDefine)
```
Comment (Jack-Works) [inline src/plugins/plugin.ts:33]: Does it work for mobile? Maybe it need a new flag like `if (Flags.poll_enabled) plugins.add(PollsPluginDefine)`

File: src/plugins/Polls/Services.ts [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+    const results = new Array<number>(options.length).fill(0)
+    const resultsObj: Object = { ...results }
+    const optionsObj: Object = { ...options }
+
+    const poll_item = {
+        ...poll,
+        results: resultsObj,
+        options: optionsObj,
+        start_time: start_time.getTime(),
```
Comment (septs) [inline src/plugins/Polls/Services.ts:41]: plz use `Date.prototype.toISOString()` (ISO 8601 formatted)

**Review by septs — CHANGES_REQUESTED** [review]

File: src/plugins/Polls/utils.ts (hunk 13 above)
Comment (Jack-Works) [inline src/plugins/Polls/utils.ts:4]: If you don't have a JSON schema(2nd parameter), don't provide it, or the metadata might be dropped

File: src/plugins/Polls/Services.ts (hunk 2 above)
Comment (Jack-Works) [inline src/plugins/Polls/Services.ts:34]: why type it as `Object`?

File: src/plugins/Polls/Services.ts [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+export type PollGunDB = PollMetaData
+
+interface voteProps {
+    poll: PollGunDB
+    index: number
+}
+
+export async function vote(props: voteProps) {
+    const { poll, index } = props
+    let results: Array<number> = [0, 0]
```
Comment (Jack-Works) [inline src/plugins/Polls/Services.ts:66]: No need to type this one. it will be infered

File: src/plugins/Polls/UI/PollsInPost.tsx [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    useEffect(() => {
+        const { key } = PollMetadataReader(props.message.meta).val as PollMetaData
+        if (key) {
+            Services.Plugin.invokePlugin('maskbook.polls', 'getPollByKey', { key }).then((res) => {
+                setUpdatedPoll(res as PollMetaData)
+            })
+        }
+    }, [])
```
Comment (Jack-Works) [inline src/plugins/Polls/UI/PollsInPost.tsx:41]: it seems like this effect will run only once. is this deliberated?

File: src/plugins/Polls/UI/Polls.tsx [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    return (
+        <Card className={classes.card} onClick={() => onClick?.()}>
+            <div className={classes.line}>
+                <div style={{ fontSize: '16px' }}>{poll.question}</div>
+                {!status || status === PollStatus.Inactive ? null : (
+                    <div className={classes.status}>
+                        {status === PollStatus.Voting ? <CircularProgress size={18} /> : null}
+                        <span className={classes.statusText}>{status}</span>
```
Comment (Jack-Works) [inline src/plugins/Polls/UI/Polls.tsx:96]: Please don't directly use `enum` value as displayed value

File: src/plugins/Polls/UI/Polls.tsx [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+                    </div>
+                )}
+            </div>
+            <div>
+                {poll.options.map((option, index) => (
+                    <div
```
Comment (Jack-Works) [inline src/plugins/Polls/UI/Polls.tsx:102]: Would you like to use https://material-ui.com/components/lists/ to rewrite this part?

File: src/plugins/Polls/UI/Polls.tsx [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+            return `Deadline: ${format(deadline, 'yyyy-MM-dd HH:mm:ss')}`
+        } else {
+            return 'sorry, cannot get correct deadline...'
+        }
+    }
+
+    return (
+        <Card className={classes.card} onClick={() => onClick?.()}>
+            <div className={classes.line}>
+                <div style={{ fontSize: '16px' }}>{poll.question}</div>
```
Comment (Jack-Works) [inline src/plugins/Polls/UI/Polls.tsx:92]: You may want to use `Typography`

File: src/plugins/Polls/UI/Polls.tsx [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+    const totalVotes = poll.results.reduce(
+        (accumulator: number, currentValue: number): number => accumulator + currentValue,
+    )
+
+    const getDeadline = (date: number) => {
+        const deadline = new Date(date)
+        if (isValid(deadline)) {
+            return `Deadline: ${format(deadline, 'yyyy-MM-dd HH:mm:ss')}`
+        } else {
+            return 'sorry, cannot get correct deadline...'
```
Comment (Jack-Works) [inline src/plugins/Polls/UI/Polls.tsx:85]: Maybe `Unknown`

File: src/plugins/Polls/UI/Polls.tsx (hunk 5 above)
Comment (Jack-Works) [inline src/plugins/Polls/UI/Polls.tsx:83]: "${time} left until vote ends." You may want to use https://mdn.io/Intl.RelativeTimeFormat to calculate that

File: src/plugins/Polls/Services.ts (hunk 3 above)
Comment (Jack-Works) [reply src/plugins/Polls/Services.ts:106]: it's the problem of `@types/gun`

File: src/plugins/Polls/UI/PollsDialog.tsx (hunk 6 above)
Comment (Jack-Works) [reply src/plugins/Polls/UI/PollsDialog.tsx:83]: IMO it's not a tuple(fixed length) based on the context.

File: src/plugins/Polls/UI/PollsDialog.tsx (hunk 7 above)
Comment (Jack-Works) [reply src/plugins/Polls/UI/PollsDialog.tsx:102]: Did you have considered using https://tc39.es/proposal-temporal/docs/?

File: src/plugins/Polls/utils.ts (hunk 13 above)
Comment (Jack-Works) [reply src/plugins/Polls/utils.ts:4]: what do you want to say?

File: src/plugins/Polls/define.tsx (hunk 14 above)
Comment (Jack-Works) [reply src/plugins/Polls/define.tsx:16]: what do you want to say?

File: src/plugins/Polls/Services.ts (hunk 15 above)
Comment (Jack-Works) [reply src/plugins/Polls/Services.ts:118]: it's the problem of `@types/gun`

File: src/plugins/Polls/Services.ts (hunk 18 above)
Comment (Jack-Works) [reply src/plugins/Polls/Services.ts:41]: this data is posted to the web IMO, does it any reason to not use UNIX timestamp?

File: src/plugins/Polls/UI/PollsDialog.tsx (hunk 7 above)
Comment (1999-XiChi) [reply src/plugins/Polls/UI/PollsDialog.tsx:102]: 发现可以用这个api：https://date-fns.org/v2.16.1/docs/add

File: src/plugins/Polls/UI/PollsInPost.tsx (hunk 20 above)
Comment (1999-XiChi) [reply src/plugins/Polls/UI/PollsInPost.tsx:41]: Yes, getPollByKey function is used to render the voting result data for the first time.

File: src/plugins/Polls/UI/Polls.tsx (hunk 5 above)
Comment (1999-XiChi) [reply src/plugins/Polls/UI/Polls.tsx:83]: good idea

Comment (1999-XiChi) [reply src/plugins/Polls/UI/Polls.tsx:83]: I want to handle it like this: When there are less than 24 hours left, the text is shown as "${time} left until vote ends.".Ohterwise, it's shown as `Deadline: ${format(deadline, 'yyyy-MM-dd HH:mm')}`

File: src/plugins/Polls/Services.ts (hunk 18 above)
Comment (1999-XiChi) [reply src/plugins/Polls/Services.ts:41]: why use ISO 8601 formatted? I'm intended to store UNIX timestamp data into GunDB

File: src/plugins/Polls/UI/Polls.tsx (hunk 5 above)
Comment (Jack-Works) [reply src/plugins/Polls/UI/Polls.tsx:83]: actually you don't need to do anything.

Comment (1999-XiChi) [reply src/plugins/Polls/UI/Polls.tsx:83]: okk

Comment (1999-XiChi) [reply src/plugins/Polls/UI/Polls.tsx:83]: image
It seems that typescript don't support this property...

Comment (1999-XiChi) [reply src/plugins/Polls/UI/Polls.tsx:83]: Maybe I can use https://date-fns.org/v2.16.1/docs/formatDistance

Comment (Jack-Works) [reply src/plugins/Polls/UI/Polls.tsx:83]: Please use `as any` as a workaround and paste this link: microsoft/TSJS-lib-generator#917

Comment (1999-XiChi) [reply src/plugins/Polls/UI/Polls.tsx:83]: With `Intl.RelativeTimeFormat.format()` I need to calculate the number of days\hours\minutes needed, why not use the `formatDistance` of date-fns directly
```typescript
            const lang = useValueRef(languageSettings)
            const localeMapping = () => {
                switch (lang) {
                    case 'en':
                        return enUS
                    case 'ja':
                        return ja
                    case 'zh':
                        return zhTW
                    default:
                        return enUS
                }
            }
            const time = formatDistance(new Date(poll.start_time), new Date(poll.end_time), {
                locale: localeMapping(),
            })
            return t('plugin_poll_deadline', { time })
```

**Comment by Tedko** [comment]
@1999-XiChi did you translate the japanese from ENG of CHN? our JP community can help the i18n

**Comment by 1999-XiChi** [comment]
> @1999-XiChi did you translate the japanese from ENG of CHN? our JP community can help the i18n

I use Google Translate to translate Japanese. How to contact our JP community？

File: src/plugins/Polls/UI/PollsDialog.tsx [hunk 25]
```diff
@@ -17,6 +17,7 @@ import {
     CircularProgress,
 } from '@material-ui/core'
 import AddIcon from '@material-ui/icons/Add'
+import { add } from 'date-fns'
```
Comment (septs) [inline src/plugins/Polls/UI/PollsDialog.tsx:20]:
```suggestion
import { add as addDate } from 'date-fns'
```

**Comment by 1999-XiChi** [comment]
How to deal with E2E and License Compliance errors？

File: src/plugins/plugin.ts (hunk 17 above)
Comment (Jack-Works) [reply src/plugins/plugin.ts:33]: I see you marked it as resolved but I don't see the code change

File: src/plugins/Polls/UI/PollsInPost.tsx [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+                setStatus(PollStatus.Voted)
+                setUpdatedPoll(res as PollMetaData)
+            })
+        } else {
+            setStatus(PollStatus.Closed)
+        }
+    }
+
+    useEffect(() => {
+        const { key } = PollMetadataReader(props.message.meta).val as PollMetaData
```
Comment (Jack-Works) [inline src/plugins/Polls/UI/PollsInPost.tsx:35]: Please don't do this. Either use `unwrap` or handle the error by `if (result.err)`

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Review by septs — CHANGES_REQUESTED** [review]
Add a `README.md` file, to `src/plugins/Polls`.

Section Structure Reference:
https://github.com/DimensionDev/Maskbook/blob/5ab3ac39715b265fbaff9629bbfc815392f4f468/src/plugins/FileService/README.md

File: src/plugins/Polls/README.md [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+
+- [x] Could send a poll or vote for options
+- [x] Show all polls user initiated
+- [x] Save voting data into gunDB
+- [ ] \[TODO\] Save voting data into Blockchain
+- [ ] \[TODO\] Show voting history of a poll
+
+## Related projects
+
+- <https://github.com/1999-XiChi/Maskbook-Poll-Contract>
```
Comment (septs) [inline src/plugins/Polls/README.md:13]: plz transfer this project to @DimensionDev.

**Review by Jack-Works — DISMISSED** [review]
LGTM, thanks for your work! 🎉 @septs and @guanbinrui, please review

**Comment by septs** [comment]
@1999-XiChi rebase master.

**Review by 1999-XiChi — COMMENTED** [review]
> @1999-XiChi rebase master.

It's done now

**Review by Jack-Works — APPROVED** [review]

**Comment by septs** [comment]
image

**Review by septs — APPROVED** [review]

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**